core/ops.rs

```diff
@@ -28,119 +25,101 @@ pub enum Op {
-      let bufs = bufs[1..].into();
-      let fut = op_fn(state.clone(), args, bufs).map(move |result| {
-        state.json_serialize_op_result(Some(promise_id), result)
-      });
-      Ok(Op::Async(Box::pin(fut)))
-    };
-
-    let base_op_fn = move |state: Rc<Self>, bufs: BufVec| -> Op {
-      match try_dispatch_op(state.clone(), bufs) {
-        Ok(op) => op,
-        Err(err) => Op::Sync(state.json_serialize_op_result(None, Err(err))),
-      }
-    };
-
-    self.register_op(name, base_op_fn)
-  }
+/// Collection for storing registered ops. The special 'get_op_catalog'
+/// op with OpId `0` is automatically added when the OpTable is created.
+pub struct OpTable(IndexMap<String, Rc<OpFn>>);
 
-  fn json_serialize_op_result(
+impl OpTable {
+  pub fn route_op(
     &self,
-    promise_id: Option<u64>,
-    result: Result<Value, ErrBox>,
-  ) -> Box<[u8]> {
-    let value = match result {
-      Ok(v) => json!({ "ok": v, "promiseId": promise_id }),
-      Err(err) => json!({
-        "promiseId": promise_id ,
-        "err": {
-          "className": self.get_error_class_name(&err),
-          "message": err.to_string(),
-        }
-      }),
-    };
-    serde_json::to_vec(&value).unwrap().into_boxed_slice()
-  }
-
-  fn get_error_class_name(&self, _err: &ErrBox) -> &'static str {
-    "Error"
+    op_id: OpId,
+    state: Rc<RefCell<OpState>>,
+    bufs: BufVec,
+  ) -> Op {
+    if op_id == 0 {
+      let ops = self.get_op_catalog();
+      let buf = serde_json::to_vec(&ops).map(Into::into).unwrap();
+      return Op::Sync(buf);
+    }
+    if let Some(op_fn) = self.get_index(op_id).map(|(_, op_fn)| op_fn.clone()) {
+      (op_fn)(state, bufs)
+    } else {
+      Op::NotFound
+    }
   }
-}
-
-/// Collection for storing registered ops. The special 'get_op_catalog'
-/// op with OpId `0` is automatically added when the OpTable is created.
-pub struct OpTable<S>(IndexMap<String, Rc<OpFn<S>>>);
 
-impl<S: OpRegistry> OpTable<S> {
   pub fn get_op_catalog(&self) -> HashMap<String, OpId> {
     self.keys().cloned().zip(0..).collect()
   }
 
-  fn op_get_op_catalog(state: Rc<S>, _bufs: BufVec) -> Op {
-    let ops = state.get_op_catalog();
-    let buf = serde_json::to_vec(&ops).map(Into::into).unwrap();
-    Op::Sync(buf)
+  pub fn register_op<F>(&mut self, name: &str, op_fn: F) -> OpId
+  where
+    F: Fn(Rc<RefCell<OpState>>, BufVec) -> Op + 'static,
+  {
+    let (op_id, prev) = self.insert_full(name.to_owned(), Rc::new(op_fn));
+    assert!(prev.is_none());
+    op_id
   }
 }
 
-impl<S: OpRegistry> Default for OpTable<S> {
+impl Default for OpTable {
   fn default() -> Self {
-    Self(
-      once(("ops".to_owned(), Rc::new(Self::op_get_op_catalog) as _)).collect(),
-    )
+    Self(once(("ops".to_owned(), Rc::new(dummy) as _)).collect())
   }
 }
 
-impl<S> Deref for OpTable<S> {
-  type Target = IndexMap<String, Rc<OpFn<S>>>;
+fn dummy(_state: Rc<RefCell<OpState>>, _v: BufVec) -> Op {
+  todo!()
+}
```

Comment:
I'm leaving this here to be courteous and not rip out IndexTable. Previous implementation was simpler https://github.com/denoland/deno/blob/v1.3.3/core/ops.rs#L72-L27

Reply:
I did see that old implementation. There was nothing wrong with it - except for that it was more lines.
Roll back if this variant gives you trouble.